app/src/behaviors/behavior_smart_interrupt.c

```diff
@@ -0,0 +1,288 @@
+    LOG_DBG("%d smart_interrupt keybind released", event.position);
+    struct active_smart_interrupt *smart_interrupt = find_smart_interrupt(event.position);
+    if (smart_interrupt == NULL) {
+        return ZMK_BEHAVIOR_OPAQUE;
+    }
+    smart_interrupt->is_pressed = false;
+    behavior_keymap_binding_released(&cfg->behaviors[1], event);
+    reset_timer(k_uptime_get(), smart_interrupt);
+    return ZMK_BEHAVIOR_OPAQUE;
+}
+
+static int behavior_smart_interrupt_init(const struct device *dev) {
+    static bool init_first_run = true;
+    if (init_first_run) {
+        for (int i = 0; i < ZMK_BHV_MAX_ACTIVE_SMART_INTERRUPTS; i++) {
+            k_work_init_delayable(&active_smart_interrupts[i].release_timer,
+                                  behavior_smart_interrupt_timer_handler);
+            clear_smart_interrupt(&active_smart_interrupts[i]);
+        }
+    }
+    init_first_run = false;
+    return 0;
+}
+
+static const struct behavior_driver_api behavior_smart_interrupt_driver_api = {
+    .binding_pressed = on_smart_interrupt_binding_pressed,
+    .binding_released = on_smart_interrupt_binding_released,
+};
+
+static int smart_interrupt_position_state_changed_listener(const zmk_event_t *eh);
+static int smart_interrupt_layer_state_changed_listener(const zmk_event_t *eh);
+
+ZMK_LISTENER(behavior_smart_interrupt, smart_interrupt_position_state_changed_listener);
+ZMK_SUBSCRIPTION(behavior_smart_interrupt, zmk_position_state_changed);
+
+ZMK_LISTENER(behavior_smart_interrupt2, smart_interrupt_layer_state_changed_listener);
+ZMK_SUBSCRIPTION(behavior_smart_interrupt2, zmk_layer_state_changed);
+
+static int smart_interrupt_position_state_changed_listener(const zmk_event_t *eh) {
+    struct zmk_position_state_changed *ev = as_zmk_position_state_changed(eh);
+    if (ev == NULL) {
+        return ZMK_EV_EVENT_BUBBLE;
+    }
+    for (int i = 0; i < ZMK_BHV_MAX_ACTIVE_SMART_INTERRUPTS; i++) {
+        struct active_smart_interrupt *smart_interrupt = &active_smart_interrupts[i];
+        if (!smart_interrupt->is_active) {
+            continue;
+        }
+        if (smart_interrupt->position == ev->position) {
+            continue;
+        }
+        if (!is_other_key_shared(smart_interrupt, ev->position)) {
+            LOG_DBG("Smart Interrupt interrupted, ending at %d %d", smart_interrupt->position,
+                    ev->position);
+            smart_interrupt->is_active = false;
+            struct zmk_behavior_binding_event event = {.position = smart_interrupt->position,
+                                                       .timestamp = k_uptime_get()};
+            if (smart_interrupt->is_pressed) {
+                behavior_keymap_binding_released(&smart_interrupt->config->behaviors[1], event);
+            }
+            behavior_keymap_binding_pressed(&smart_interrupt->config->behaviors[2], event);
+            behavior_keymap_binding_released(&smart_interrupt->config->behaviors[2], event);
```

Comment:
Ditto on timing, and I would wrap all "trigger the interrupt behavior" into a shared function.